File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        mat = [[None for i in range(n)] for _ in range(n)]
        
        left, right = 0, n
        top, bottom = 0, n
        val = 1
        while left < right and top < bottom:
            # put value in the top row
            for i in range(left, right):
                mat[top][i] = val
                val += 1
            top += 1
            
            # put value in the right col
            for i in range(top, bottom):
                mat[i][right-1] = val
                val += 1
            right -= 1
            
            # if not (left < right and top < bottom):
            #     break
            # put value in the bottom row
            for i in range(right-1, left-1, -1):
                mat[bottom-1][i] = val
                val += 1
            bottom -= 1
            
            # put value in the left col
            for i in range(bottom-1, top-1, -1):
                mat[i][left] = val
                val += 1
            left += 1
        return mat

File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int left = 0;
        int right = n;
        int top = 0;
        int bottom = n;
        int value = 1;

        vector<vector<int>> res(n, vector<int> (n));
        
        while(left < right && top < bottom){
            for(int i=left; i<right; i++){
                res[top][i] = value;
                value++;
            }
            top++;

            for(int i=top; i<bottom; i++){
                res[i][right-1] = value;
                value++;
            }
            right--;

            for(int i=right-1; i>=left; i--){
                res[bottom-1][i] = value;
                value++;
            }
            bottom--;

            for(int i=bottom-1; i>=top; i--){
                res[i][left] = value;
                value++;
            }
            left++;
        }
        return res;
    }
};